asserts/tool/sign.go

```diff
@@ -0,0 +1,249 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package tool offers tooling to sign assertions.
+package tool
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/snapcore/snapd/asserts"
+)
+
+// The supported media types for the input of assertion signing.
+const (
+	JSONInput = "application/json"
+	YAMLInput = "application/x-yaml"
+)
+
+// SignRequest specifies the complete input for signing an assertion.
+type SignRequest struct {
+	// The key to use can be speficied either passing the text of
+	// an account-key assertion in AccountKey
+	AccountKey []byte
+	// or passing the key id in KeyID
+	KeyID string
+	// and an optional account-id of the signer (if left out headers value are consulted) in AuthorityID
+	AuthorityID string
+
+	// The assertion type (as a string)
+	AssertionType string
+	// StatementMediaType specifies the media type of the input
+	StatementMediaType string
+	// Statement is used as input to construct the assertion
+	// it's a mapping encoded as either JSON or YAML
+	// (specified in StatementMediaType)
+	// either of just the header fields of the assertion
+	// or containting exactly two entries
+	// "headers": mapping with the header fields
+	// "body": used as the body of the assertion
+	Statement []byte
+
+	// The revision of the new assertion
+	Revision int
+}
+
+func parseStatement(req *SignRequest, dest interface{}) error {
+	switch req.StatementMediaType {
+	case YAMLInput:
+		err := yaml.Unmarshal(req.Statement, dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as YAML: %v", err)
+		}
+	case JSONInput:
+		dec := json.NewDecoder(bytes.NewBuffer(req.Statement))
+		// we want control over supporting only integers
+		dec.UseNumber()
+		err := dec.Decode(dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as JSON: %v", err)
+		}
+	default:
+		return fmt.Errorf("unsupported media type for assertion input: %q", req.StatementMediaType)
+	}
+	return nil
+}
+
+type nestedStatement struct {
+	Headers map[string]interface{} `yaml:"headers" json:"headers"`
+	Body    string                 `yaml:"body" json:"body"`
+}
+
+// Sign produces the text of a signed assertion as specified by req.
+func Sign(req *SignRequest, keypairMgr asserts.KeypairManager) ([]byte, error) {
+	typ := asserts.Type(req.AssertionType)
+	if typ == nil {
+		return nil, fmt.Errorf("invalid assertion type: %q", req.AssertionType)
+	}
+	if req.Revision < 0 {
+		return nil, fmt.Errorf("assertion revision cannot be negative")
+	}
+	if req.AccountKey == nil && req.KeyID == "" {
+		return nil, fmt.Errorf("both account-key and key id were not specified")
+	}
+
+	var nestedStatement nestedStatement
+	err := parseStatement(req, &nestedStatement)
+	if err != nil {
+		return nil, err
+	}
+	if nestedStatement.Headers == nil {
+		// flat headers, reparse
```

Comment:
I'm picking this point to comment because it's a good example to highlight, but this is not specific to it: the initial gut reaction while reading this is that the tool is quite loose, with excessive flexibility and fiddling. This is not a good property for something dealing with assertions and signatures.

This ought to be much more tight: take a set of key/value pairs, and that's it. Fields are either there, with their proper names, or they're not. The body will need special handling, but that's trivial to get right by simply saying we have a "body" field, which we should never have as an actual header anyway as it would be very confusing.

That means this interface ends up looking a lot like Database.Sign, which is a good indication and makes me wonder if we need this package at all. Let's see how much we can take out of it.